### PR TITLE
[MCKIN-7023] Allows translated strings in XBlock Django templates

### PR DESCRIPTION
**Pull request by pomegranited**
Adds `i18n` support to XBlock Django templates by adding support for `trans` and `blocktrans` template tags during template rendering.

The XBlock translations are provided by the `i18n` runtime service, which must be passed to `ResourceLoader.render_django_template`.

In order for an XBlock's template strings to be translated, the XBlock must:

* Require the `i18n` service, see e.g. [xblock-drag-and-drop-v2](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/26a8c5c5edb65beae24707353432654b50e7a230/drag_and_drop_v2/drag_and_drop_v2.py#L34)
   ```python
   @XBlock.needs('i18n')
   class MyXBlock(XBlock):
   ...
   ```
* Render templates using [`ResourceLoader.render_django_template`](https://github.com/open-craft/xblock-utils/blob/MCKIN-7023-template-i18n/xblockutils/resources.py#L50), passing the `i18n` service,  e.g.:
   ```python
   def studio_view(self, context):
      fragment = Fragment()
      fragment.add_content(loader.render_django_template('/templates/html/studio_edit.html', 
                                                         i18n_service=self.runtime.service(self, 'i18n'))
   ```
* In the HTML template, [load the `i18n` template tags](https://github.com/edx-solutions/xblock-drag-and-drop-v2/blob/dd91427874637b8abad7fdeacaf45dfb50afaee1/drag_and_drop_v2/templates/html/drag_and_drop_edit.html#L1):
   ```html
   {% load i18n %}
   ```
* Mark translatable strings using the [`trans`](https://docs.djangoproject.com/en/2.0/topics/i18n/translation/#trans-template-tag) and [`blocktrans`](https://docs.djangoproject.com/en/2.0/topics/i18n/translation/#blocktrans-template-tag) tags, as described in the Django documentation.
* Extract and generate the translation `.po` and `.mo` files in your XBlock's `translations` directory, as described in the [XBlock tutorial](http://edx.readthedocs.io/projects/xblock-tutorial/en/latest/edx_platform/edx_lms.html#internationalization-support).

**JIRA tickets**: [MCKIN-7023](https://edx-wiki.atlassian.net/browse/MCKIN-7023), [OSPR-2279](https://openedx.atlassian.net/browse/OSPR-2279)

**Discussions**:
* [WL-230](https://openedx.atlassian.net/browse/WL-230)
* [YONK-879](https://openedx.atlassian.net/browse/YONK-879?focusedCommentId=307771&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-307771)
* [edx-code mailing list](https://groups.google.com/forum/#!searchin/edx-code/xblock$20i18n%7Csort:date/edx-code/XoIQwvIMi4c/rwlyk76aAwAJ)
* [openedx-translation mailing list]( https://groups.google.com/d/msg/openedx-translation/cLR5tZI5oqQ/Eln357QzAwAJ)

**Dependency of:**
* https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/154
* https://github.com/edx-solutions/xblock-image-explorer/pull/37
* https://github.com/edx/edx-platform/pull/17670

**Screenshots**: See https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/154 for before and after screenshots using dummy translations.

**Sandbox URL**:  sandbox is being provisioned.

* LMS: https://pr17670.sandbox.opencraft.hosting/
* Studio: https://studio-pr17670.sandbox.opencraft.hosting/

**Merge deadline**: ASAP

**Testing instructions**:

See https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/154 for test instructions.

**Reviewers**
- [ ] @mtyaka 
- [ ] edX reviewer[s] TBD

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @pomegranited! I've created [OSPR-2279](https://openedx.atlassian.net/browse/OSPR-2279) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/xblock-utils/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by nedbat** [comment]
Ping me when you are ready.

File: tox.ini [hunk 1]
```diff
@@ -8,7 +8,7 @@ envlist = py{27}-django{18,111}
 deps =
     -rrequirements.txt
     -rtest_requirements.txt
-    -egit+https://github.com/edx/xblock-sdk.git#egg=xblock-sdk
+    -egit+https://github.com/edx/xblock-sdk.git@48cb0d5066a1f5ab8eecba41bb3d63c78c985a81#egg=xblock-sdk
```
Comment (pomegranited) [inline tox.ini]: Something about https://github.com/edx/xblock-utils/pull/45 caused tox to return this error, so I've reverted to the revision just prior.
```
django18 runtests: commands[3] | python ./run_tests.py --with-coverage --cover-package=xblockutils
usage: run_tests.py test [-h] [--version] [-v {0,1,2,3}] [--settings SETTINGS]
                         [--pythonpath PYTHONPATH] [--traceback] [--no-color]
                         [--noinput] [--failfast] [--testrunner TESTRUNNER]
                         [--liveserver LIVESERVER] [-t TOP_LEVEL] [-p PATTERN]
                         [-k] [-r] [-d]
                         [test_label [test_label ...]]
run_tests.py test: error: unrecognized arguments: --with-coverage --cover-package=xblockutils
ERROR: InvocationError: '/home/travis/build/edx/xblock-utils/.tox/django18/bin/python ./run_tests.py --with-coverage --cover-package=xblockutils'
```

Some version incompatibility introduced with `nose`, `coverage`, or other?

File: xblockutils/resources.py [hunk 2]
```diff
@@ -51,8 +52,24 @@ def render_django_template(self, template_path, context=None):
         Evaluate a django template by resource path, applying the provided context
         """
         context = context or {}
+        libraries = {
+            'i18n': 'xblockutils.templatetags.i18n',
+        }
+
+        # If django>1.8, Engine can load the extra templatetag libraries
+        if django.VERSION[0] > 0 and django.VERSION[1] > 8:
```
Comment (mtyaka) [inline xblockutils/resources.py]: This does not work correctly for `2.0`.

File: xblockutils/resources.py [hunk 3]
```diff
@@ -51,8 +52,24 @@ def render_django_template(self, template_path, context=None):
...
+            engine = Engine(libraries=libraries)
+        else:
+            # For django 1.8, we have to load the libraries manually
+            from django.template import base as TemplateBase
+            for library_name in libraries:
+                library = TemplateBase.import_library(libraries[library_name])
+                if library:
+                    TemplateBase.libraries[library_name] = library
```
Comment (mtyaka) [inline xblockutils/resources.py]: This is changing global state in `TemplateBase` module. Would it make sense to we revert the changes made to `TemplateBase.libraries` after we are done rendering the template, so that these changes don't affect template rendering in unrelated parts of the codebase?

File: xblockutils/templatetags/i18n.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    def render(self, context):
+        """
+        Renders the translated text using the XBlock i18n service, if available.
+        """
+        django_translated = self.do_translate.render(context)
+
+        i18n_service = context.get('_i18n_service', None)
+        if i18n_service:
+            return i18n_service.gettext(django_translated)
```
Comment (mtyaka) [inline xblockutils/templatetags/i18n.py]: I guess this does not support things such as `{% trans "my string" as variable %}` or [`blocktrans` features](https://docs.djangoproject.com/en/2.0/topics/i18n/translation/#blocktrans-template-tag) such as variable substitution or support for plural forms. I'm not sure how easy it would be to support these, though. If it's hard, I think a proxy like this is fine for the initial implementation.

Comment (pomegranited) [reply xblockutils/templatetags/i18n.py]: Oo good point -- I dug a bit deeper and found a way to do that which should continue to work through Django 2.0. See b9c2ae2.

**Comment by pomegranited** [comment]
@nedbat Could I get a review ASAP on this approach, even if it's not a full code review?  We'd like to use this into edx-solutions soon, but don't want to stray from the upstream path.

**Comment by nedbat** [comment]
@edx/platform-team @cpennington @nasthagiri Does someone want to take a look at this?

**Comment by nasthagiri** [comment]
Perhaps @bderusha or @amangano-edx ?

File: xblockutils/templatetags/i18n.py [hunk 5]
```diff
@@ -44,6 +45,8 @@ def merge_translation(self, context):
         # Revert to original translation object
         if language in self._translations:
             trans_real._translations[language] = self._translations[language]
+            # Re-activate the current language to resets translation caches
+            trans_real.activate(language)
```
Comment (pomegranited) [inline xblockutils/templatetags/i18n.py]: @mtyaka I found a bug with my latest approach -- the translation change wasn't getting reverted completely, and so the XBlock translations could bleed into other places in the platform.  

This line ensures that [`trans_real._active.value`](https://github.com/django/django/blob/1.8.19/django/utils/translation/trans_real.py#L213-L225) is reverted too.

Updated the tests to move the "untranslated text" check into the same method as the "translated text" check, which demonstrates that the translation was fully reverted.

File: xblockutils/templatetags/i18n.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+            translation = trans_real.translation(language)
+            translation.merge(i18n_service)
+
+        yield
+
+        # Revert to original translation object
+        if language in self._translations:
+            trans_real._translations[language] = self._translations[language]
+            # Re-activate the current language to resets translation caches
```
Comment (mtyaka) [inline xblockutils/templatetags/i18n.py]: Nit: typo resets -> reset

**Comment by mtyaka** [comment]
:+1:

- [x] I tested this: I verified that the code works correctly with i18n aware DnDv2 and image-explorer xblocks.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~ n/a
- [ ] Includes documentation

**Comment by pomegranited** [comment]
@nedbat  We've done our internal review, so this is ready for your team's review now.
CC @bderusha @amangano-edx

File: xblockutils/templatetags/i18n.py (hunk 6 above)
Comment (pomegranited) [reply xblockutils/templatetags/i18n.py]: Fixed with 8c81c05

**Comment by pomegranited** [comment]
@bderusha @amangano-edx CC @nasthagiri @nedbat @bradenmacdonald   We'd appreciate a review of the approach used here, so we can move ahead with it on the client's fork while we wait for a full review.

Is this possible to have ASAP?  Thank you!

**Comment by bderusha** [comment]
@pomegranited Anthony and I will not be able to review this anytime soon.  I think Ned's tag of the @edx/platform-team is a better bet.

**Comment by pomegranited** [comment]
Thanks for your reply @bderusha, that's good to know.

@nedbat Could you please pursue this with @edx/platform-team members?  As I mentioned before, what we need ASAP is a review of the general approach, then we can sort the details as part of a detailed review with less urgency. CC @bradenmacdonald

**Comment by cpennington** [comment]
The approach looks reasonable to me.

**Comment by pomegranited** [comment]
Thank you @cpennington !

**Comment by pomegranited** [comment]
Squashed  c868900ffdec380ea1a02cca3826f945af18fafe...8c81c056a90bef522837f7c26ed5af3e1122fd9c into ba54084 so I can use this change on the client's fork now.

**Comment by pomegranited** [comment]
@cpennington CC @edx/platform-team @nedbat  Do you have an estimated date for when this will be reviewed?  Thank you!

**Comment by pomegranited** [comment]
@cpennington @nedbat -- We're working on allowing XBlocks to use translated text in Javascript, and would appreciate your review of our approach.  Would you be willing to have a look?  Or is there anyone on the frontend team who is working on i18n?

**Comment by pomegranited** [comment]
Hi @adamstankiewicz -- do you know when this PR will be reviewed?  It's getting complicated to manage in our client branches, and affects a lot of other open PRs, so we'd like to get it in ASAP. CC @bradenmacdonald

**Comment by adamstankiewicz** [comment]
@pomegranited  I think I was assigned to this PR by mistake as I have no context around XBlocks or what this PR is for. I'll work with @mduboseedx to figure out who would be a good person to review.

**Comment by mduboseedx** [comment]
@edx/platform-team @macdiesel Brian, could you have someone on your team review this?

**Comment by pomegranited** [comment]
Thank you for sorting this out @adamstankiewicz @mduboseedx @macdiesel !

**Comment by edx-webhook** [comment]
@pomegranited 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/xblock-utils&entry.867055334=https://github.com/edx/xblock-utils/pull/48&entry.1484655318=https://github.com/pomegranited&entry.752974735=2018-03-06+23:08&entry.1917517419=2018-05-07+17:46&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by pomegranited** [comment]
Woo hoo!  Thank you @macdiesel !

**Comment by pomegranited** [comment]
@macdiesel Can you trigger a new pypi build to release v1.1.0?

**Comment by pomegranited** [comment]
@macdiesel  Please disregard my previous message -- I was able to push a new tag for `v1.1.0`.
